language/vm/vm-runtime/src/move_vm.rs

```diff
@@ -72,17 +90,148 @@ impl VMExecutor for MoveVM {
+                        process_block_metadata(block_metadata, runtime, state_view, &mut data_cache)
+                    }))
+                }
+                // TODO: Implement the logic for processing writeset transactions.
+                TransactionBlock::WriteSet(_) => unimplemented!(""),
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+/// Transactions divided by transaction flow.
+/// Transaction flows are different across different types of transactions.
+pub enum TransactionBlock {
+    UserTransaction(Vec<SignedTransaction>),
+    WriteSet(WriteSet),
+    BlockPrologue(BlockMetadata),
+}
+
+pub fn chunk_block_transactions(txns: Vec<Transaction>) -> Vec<TransactionBlock> {
+    let mut blocks = vec![];
+    let mut buf = vec![];
+    for txn in txns {
+        match txn {
+            Transaction::BlockMetadata(data) => {
+                if !buf.is_empty() {
+                    blocks.push(TransactionBlock::UserTransaction(buf));
+                    buf = vec![];
+                }
+                blocks.push(TransactionBlock::BlockPrologue(data));
+            }
+            Transaction::WriteSet(ws) => {
+                if !buf.is_empty() {
+                    blocks.push(TransactionBlock::UserTransaction(buf));
+                    buf = vec![];
+                }
+                blocks.push(TransactionBlock::WriteSet(ws));
+            }
+            Transaction::UserTransaction(txn) => {
+                buf.push(txn);
+            }
+        }
+    }
+    if !buf.is_empty() {
+        blocks.push(TransactionBlock::UserTransaction(buf));
     }
+    blocks
+}
+
+/// Determine if a transaction is valid. Will return `None` if the transaction is accepted,
+/// `Some(Err)` if the VM rejects it, with `Err` as an error code. We verify the following
+/// items:
+/// 1. The signature on the `SignedTransaction` matches the public key included in the
+///    transaction
+/// 2. The script to be executed is in the whitelist.
+/// 3. Invokes `LibraAccount.prologue`, which checks properties such as the transaction has the
+/// right sequence number and the sender has enough balance to pay for the gas. 4.
+/// Transaction arguments matches the main function's type signature. 5. Script and modules
+/// in the transaction pass the bytecode static verifier.
+///
+/// Note: In the future. we may defer these checks to a later pass, as all the scripts we will
+/// execute are pre-verified scripts. And bytecode verification is expensive. Thus whether we
+/// want to perform this check here remains unknown.
+pub fn verify_transaction(
+    runtime: &VMRuntime,
+    txn: SignedTransaction,
+    state_view: &dyn StateView,
+) -> Option<VMStatus> {
+    trace!("[VM] Verify transaction: {:?}", txn);
+    let mut data_cache = BlockDataCache::new(state_view);
+
+    // If we fail to load the gas schedule, then we fail to process the block.
+    let gas_schedule = match runtime.load_gas_schedule(&mut data_cache, state_view) {
```

Comment:
I'll let @tzakian comment on that. I think you are correct that the transaction flow needs a bit of a review. But that work is coming so I think we should keep those issue in mind. I am not sure there is a TODO to add here, let me know otherwise